ui-admin/src/forms/designer/questions/ChoicesList.tsx

```diff
@@ -0,0 +1,158 @@
+                        ...question.choices.slice(0, i),
+                        { ...question.choices[i], text: value },
+                        ...question.choices.slice(i + 1)
+                      ]
+                    })
+                  }}
+                />
+                <TextInput
+                  disabled={readOnly}
+                  label="Value"
+                  labelClassname="mb-0"
+                  value={choice.value}
+                  onChange={value => {
+                    onChange({
+                      ...question,
+                      choices: [
+                        ...question.choices.slice(0, i),
+                        { ...question.choices[i], value },
+                        ...question.choices.slice(i + 1)
+                      ]
+                    })
+                  }}
+                />
+              </div>
+
+              <div className="d-flex flex-column justify-content-between ms-2">
+                <IconButton
+                  aria-label="Move this choice before the previous one"
+                  disabled={readOnly || i === 0}
+                  icon={faChevronUp}
+                  variant="light"
+                  onClick={() => {
+                    onChange({
+                      ...question,
+                      choices: [
+                        ...question.choices.slice(0, i - 1),
+                        question.choices[i],
+                        question.choices[i - 1],
+                        ...question.choices.slice(i + 1)
+                      ]
+                    })
+                  }}
+                />
+
+                <IconButton
+                  aria-label="Move this choice after the next one"
+                  disabled={readOnly || i === question.choices.length - 1}
+                  icon={faChevronDown}
+                  variant="light"
+                  className="btn btn-light"
+                  onClick={() => {
+                    onChange({
+                      ...question,
+                      choices: [
+                        ...question.choices.slice(0, i),
+                        question.choices[i + 1],
+                        question.choices[i],
+                        ...question.choices.slice(i + 2)
+                      ]
+                    })
+                  }}
+                />
+              </div>
+
+              <div className="d-flex flex-column justify-content-between flex-shrink-0 ms-2">
+                <IconButton
+                  aria-label="Delete this choice"
+                  disabled={readOnly}
+                  icon={faTimes}
+                  variant="light"
+                  onClick={() => {
+                    onChange({
+                      ...question,
+                      choices: [
+                        ...question.choices.slice(0, i),
+                        ...question.choices.slice(i + 1)
+                      ]
+                    })
```

Comment:
Yeah, seems like we have different preferred styles for when to pull things out into component-level functions.  Let me know your thoughts -- I lean towards pulling most non-trivial things into component-level functions because it (to me) makes the markup of the HTML elements easier to follow.  Let me know what your approach is (and to what extent there is a general DSP guideline on when to inline)

Reply:
FWIW, each component on its own is small enough to be digestible to me. Extracted functions could still be useful, as you pointed out elsewhere, if naming the concept would be useful to understanding the code... or to provide a testable unit, which I think is a better reason in this specific case.

Reply:
Since all these buttons do different things, I'm not sure extracting functions would be any less verbose / easier to understand. As is, the data manipulation is right next to the button label describing its intent.